src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs

```diff
@@ -1284,15 +1288,15 @@ private ITryStatement CreateBoundTryStatementOperation(BoundTryStatement boundTr
 
         private ICatchClause CreateBoundCatchBlockOperation(BoundCatchBlock boundCatchBlock)
         {
-            Lazy<IBlockStatement> handler = new Lazy<IBlockStatement>(() => (IBlockStatement)Create(boundCatchBlock.Body));
-            ITypeSymbol caughtType = boundCatchBlock.ExceptionTypeOpt;
+            Lazy<IOperation> expressionDeclarationOrExpression = new Lazy<IOperation>(() => boundCatchBlock.ExceptionSourceOpt != null ? Create(boundCatchBlock.ExceptionSourceOpt) : null);
+            ITypeSymbol exceptionType = boundCatchBlock.ExceptionTypeOpt ?? ((CSharpCompilation)_semanticModel.Compilation).GetWellKnownType(WellKnownType.System_Exception);
             Lazy<IOperation> filter = new Lazy<IOperation>(() => Create(boundCatchBlock.ExceptionFilterOpt));
-            ILocalSymbol exceptionLocal = (boundCatchBlock.Locals.FirstOrDefault()?.DeclarationKind == CSharp.Symbols.LocalDeclarationKind.CatchVariable) ? boundCatchBlock.Locals.FirstOrDefault() : null;
+            Lazy<IBlockStatement> handler = new Lazy<IBlockStatement>(() => (IBlockStatement)Create(boundCatchBlock.Body));
             SyntaxNode syntax = boundCatchBlock.Syntax;
             ITypeSymbol type = null;
             Optional<object> constantValue = default(Optional<object>);
             bool isImplicit = boundCatchBlock.WasCompilerGenerated;
-            return new LazyCatchClause(handler, caughtType, filter, exceptionLocal, _semanticModel, syntax, type, constantValue, isImplicit);
+            return new LazyCatchClause(expressionDeclarationOrExpression, exceptionType, filter, handler, _semanticModel, syntax, type, constantValue, isImplicit);
```

Comment:
>return new LazyCatchClause(expressionDeclarationOrExpression, exceptionType, filter, handler, _semanticModel, syntax, type, constantValue, isImplicit); [](start = 12, length = 151)

It doesn't look like we are exposing ```boundCatchBlock.Locals```.

Reply:
These exposed through the underlying IBlockStatement of the handler. See [https://github.com/dotnet/roslyn/pull/22210/files#diff-86f74ac187f645e9c3f76409a6fac779R467](https://github.com/dotnet/roslyn/pull/22210/files#diff-86f74ac187f645e9c3f76409a6fac779R467)

---
In reply to: [141126304](https://github.com/dotnet/roslyn/pull/22210#discussion_r141126304) [](ancestors = 141126304)

Reply:
> These exposed through the underlying IBlockStatement of the handler. See [https://github.com/dotnet/roslyn/pull/22210/files#diff-86f74ac187f645e9c3f76409a6fac779R467](https://github.com/dotnet/roslyn/pull/22210/files#diff-86f74ac187f645e9c3f76409a6fac779R467)

I don't think these are the locals I am talking about. I am talking about the exception local and the locals declared in the filter. 

---
In reply to: [141348394](https://github.com/dotnet/roslyn/pull/22210#discussion_r141348394) [](ancestors = 141348394,141126304)

Reply:
Ah, did not realize that. I had filed [https://github.com/dotnet/roslyn/issues/22301](https://github.com/dotnet/roslyn/issues/22301) to track this work, but the API is straightforward so I have added it to this PR.

---
In reply to: [141764536](https://github.com/dotnet/roslyn/pull/22210#discussion_r141764536) [](ancestors = 141764536,141348394,141126304)